avr-hal-generic/src/ep.rs

```diff
@@ -0,0 +1,247 @@
+
+#[macro_export]
+macro_rules! impl_eeprom_traditional {
```

Comment:
Your implementation doesn't really match what we're doing everywhere else in `avr-hal`.  There are two main problems:

1. You are defining the "driver" type as part of the macro, so it will only exist in the HAL crates.  We've moved most modules to a design where the "driver" type is generic and defined in `avr-hal-generic` instead.  This allows writing generic code for AVR microcontrollers which will work with any supported chip.
2. You completely sidestepped the ownership model by using the `unsafe { EEPROM::ptr() }` everywhere.  Please don't do this.  Instead, the driver should consume the EEPROM register block at initialization.  This naturally enforces correct synchronization and means you never need any critical sections inside your driver code.

To give you an idea of what the module should look like, here is a rough sketch of how the API design of `avr-hal` can be applied for the EEPROM module:

##### `avr-hal-generic/src/eeprom.rs`
```rust
use core::marker;

#[derive(uDebug, Debug)]
pub struct OutOfBoundsError;

pub trait EepromOps<H> {
    const CAPACITY: u16;

    /// TODO
    ///
    /// **Warning**: This is a low-level method and should not be called directly from user code.
    fn raw_read_byte(&self, offset: u16) -> u8;
    /// TODO
    ///
    /// **Warning**: This is a low-level method and should not be called directly from user code.
    fn raw_write_byte(&mut self, offset: u16, data: u8);
    /// TODO
    ///
    /// **Warning**: This is a low-level method and should not be called directly from user code.
    fn raw_erase_byte(&mut self, offset: u16);
}

pub struct Eeprom<H, EEPROM> {
    p: EEPROM,
    _h: marker::PhantomData<H>,
}

impl<H, EEPROM> Eeprom<H, EEPROM>
where
    EEPROM: EepromOps<H>,
{
    pub fn new(p: EEPROM) -> Self {
        Self { p }
    }

    pub fn capacity(&self) -> u16 {
        EEPROM::CAPACITY
    }

    #[inline]
    pub fn read_byte(&self, offset: u16) -> u8 {
        assert!(offset < EEPROM::CAPACITY);
        self.p.raw_read_byte(offset)
    }

    #[inline]
    pub fn write_byte(&mut self, offset: u16, data: u8) {
        assert!(offset < EEPROM::CAPACITY);
        self.p.raw_write_byte(offset, data)
    }

    #[inline]
    pub fn erase_byte(&mut self, offset: u16) {
        assert!(offset < EEPROM::CAPACITY);
        self.p.raw_erase_byte(offset)
    }

    pub fn read(&self, offset: u16, buf: &mut [u8]) -> Result<(), OutOfBoundsError> {
        if buf.len() as u16 + offset > EEPROM::CAPACITY {
            return OutOfBoundsError;
        }
        for (i, byte) in buf.iter_mut().enumerate() {
            *byte = self.p.raw_read_byte(offset + i);
        }
        Ok(())
    }

    pub fn write(&mut self, offset: u16, buf: &[u8]) -> Result<(), OutOfBoundsError> {
        todo!()
    }

    pub fn erase(&mut self, from: u16, to: u16) -> Result<(), OutOfBoundsError> {
        todo!()
    }
}

impl<H, EEPROM> embedded_storage::nor_flash::ReadNorFlash for Eeprom<H, EEPROM>
where
    EEPROM: EepromOps<H>,
{
    type Error = OutOfBoundsError;
    const READ_SIZE: usize = 1;

    fn read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), Self::Error> {
        self.read(offset, bytes)
    }
}

// same for NorFlash

#[macro_export]
macro_rules! impl_eeprom_atmega {
    (
        hal: $HAL:ty,
        peripheral: $EEPROM:ty,
        capacity: $capacity:literal,
        set_address: |$periph_var:ident, $address:ident| $set_address:block,
    ) => {
        mod helper {
            pub fn wait_read(regs: &$EEPROM) {
                while regs.eecr.read().eepe().bit_is_set() {  }
            }

            pub fn set_address(regs: &$EEPROM, address: u16) {
                wait_read(regs);
                let $periph_var = regs;
                let $address = address;
                $set_address
            }

            pub fn set_erasewrite_mode(regs: &$EEPROM) {
                regs.eecr.write(|w| {
                    w.eempe().set_bit().eepm().val_0x00()
                })
            }

            // ...
        }
        impl $crate::eeprom::EepromOps<$HAL> for $EEPROM {
            const CAPACITY = $capacity;

            #[inline]
            fn raw_read_byte(&self, offset: u16) -> u8 {
                helper::set_address(self, offset);
                self.eecr.write(|w| w.eere().set_bit())
                self.eedr.read().bits()
            }

            // ...
        }
    }
}

// same for attiny EEPROM macro
```

##### `mcu/atmega-hal/src/eeprom.rs`
```rust
use avr_hal_generic::eeprom::{OutOfBoundsError, EepromOps};

pub type Eeprom = avr_hal_generic::eeprom::Eeprom<crate::Atmega, crate::pac::EEPROM>;

#[cfg(any(
    feature = "atmega328pb",
    feature = "atmega328p",
    feature = "atmega32u4"
))]
avr_hal_generic::impl_eeprom_atmega! {
    hal: crate::Atmega,
    peripheral: crate::pac::EEPROM,
    capacity: 1024,
    set_address: |peripheral, address| {
        peripheral.eear.write(|w| w.bits(address));
    },
}
```

##### `examples/arduino-uno/src/bin/uno-eeprom.rs`
```rust
// ...

fn main() -> ! {
    let dp = arduino_hal::Peripherals::take().unwrap();
    let pins = arduino_hal::pins!(dp);
    let mut serial = arduino_hal::default_serial!(dp, pins, 57600);

    let mut ep = arduino_hal::Eeprom::new(dp.EEPROM);
    let ep_capacity = ep.capacity();
    ufmt::uwriteln!(&mut serial, "eeprom capacity is:{}\r", ep_capacity).void_unwrap();

    todo!()
}
```

I hope this helps... If you need assistance, please let me know!

Reply:
@Rahix, Thank you for your tips and patience. pls review new codes.